source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -14,50 +14,121 @@
+          [this]() { runHighWatermarkCallbacks(); }) {}
+
 void EnvoyQuicServerStream::encode100ContinueHeaders(const Http::HeaderMap& headers) {
   ASSERT(headers.Status()->value() == "100");
   encodeHeaders(headers, false);
 }
-void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& /*headers*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& headers, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeHeaders (end_stream={}) {}.", *this, end_stream, headers);
+  WriteHeaders(envoyHeadersToSpdyHeaderBlock(headers), end_stream, nullptr);
+  local_end_stream_ = end_stream;
 }
-void EnvoyQuicServerStream::encodeData(Buffer::Instance& /*data*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeData (end_stream={}) of {} bytes.", *this, end_stream,
+                   data.length());
+  local_end_stream_ = end_stream;
+  // This is counting not serialized bytes in the send buffer.
+  uint64_t bytes_to_send_old = BufferedDataBytes();
+  // QUIC stream must take all.
+  WriteBodySlices(quic::QuicMemSliceSpan(quic::QuicMemSliceSpanImpl(data)), end_stream);
+  ASSERT(data.length() == 0);
+
+  uint64_t bytes_to_send_new = BufferedDataBytes();
+  ASSERT(bytes_to_send_old <= bytes_to_send_new);
+  maybeCheckWatermark(bytes_to_send_old, bytes_to_send_new,
+                      dynamic_cast<QuicFilterManagerConnectionImpl&>(*session()));
 }
-void EnvoyQuicServerStream::encodeTrailers(const Http::HeaderMap& /*trailers*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeTrailers(const Http::HeaderMap& trailers) {
+  ASSERT(!local_end_stream_);
+  local_end_stream_ = true;
+  ENVOY_STREAM_LOG(debug, "encodeTrailers: {}.", *this, trailers);
+  WriteTrailers(envoyHeadersToSpdyHeaderBlock(trailers), nullptr);
 }
+
 void EnvoyQuicServerStream::encodeMetadata(const Http::MetadataMapVector& /*metadata_map_vector*/) {
+  // Metadata Frame is not supported in QUIC.
   NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
 }
 
-void EnvoyQuicServerStream::resetStream(Http::StreamResetReason /*reason*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+void EnvoyQuicServerStream::resetStream(Http::StreamResetReason reason) {
+  // Higher layers expect calling resetStream() to immediately raise reset callbacks.
+  runResetCallbacks(reason);
+  if (local_end_stream_ && !reading_stopped()) {
+    // This is after 200 early response. Reset with QUIC_STREAM_NO_ERROR instead
+    // of propagating original reset reason.
+    StopReading();
+  } else {
+    Reset(envoyResetReasonToQuicRstError(reason));
+  }
 }
 
-void EnvoyQuicServerStream::readDisable(bool /*disable*/) { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
+void EnvoyQuicServerStream::switchStreamBlockState(bool should_block) {
+  ASSERT(FinishedReadingHeaders(),
```

Comment:
It certainly sounds unlikely, but if we can handle block before finishing headers, we should not ASSERT, a warning log seems not appropriate to me.
Also, can we add some information to the error message? Like the upstream buffer limit and bytes read for headers.

Reply:
SG